src/bufr/BufrParser/Query/DataProvider.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * (C) Copyright 2022 NOAA/NWS/NCEP/EMC
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <memory>
+
+#include <gsl/gsl-lite.hpp>
+
+namespace Ingester{
+namespace bufr {
+    typedef unsigned int FortranIdx;
+
+    enum class Typ
+    {
+        Subset,
+        DelayedRep,
+        FixedRep,
+        DelayedRepStacked,
+        DelayedBinary,
+        Sequence,
+        Repeat,
+        StackedRepeat,
+        Number,
+        Character
+    };
+
+    /// \brief Responsible for exposing the data found in a BUFR file in a C friendly way.
+    class DataProvider
+    {
+     public:
+        DataProvider() = default;
+        ~DataProvider() = default;
+
+        /// \brief Read the data from the BUFR interface for the current subset and reset the
+        /// internal data structures.
```

Comment:
So when you read BUFR data with NCEPLIB-bufr you always read the data one subset at a time. So you will have some kind of loop (see File.cpp::execute) that loops through each message and subset and calls `updateData`. updateData causes the DataProvider to load the data from the current subset into its internal data structures so that the data returned will always be for the subset we are currently on.

```cxx 
        auto dataProvider = DataProvider();  //Create the data provider
        auto resultSet = ResultSet(querySet.names());
        auto query = Query(querySet, resultSet, dataProvider);

        while (ireadmg_f(fileUnit_, subset, &iddate, SubsetLen) == 0)
        {
            while (ireadsb_f(fileUnit_) == 0)
            {
                status_f(fileUnit_, &bufrLoc, &il, &im);
                dataProvider.updateData(bufrLoc);  // Read the data for the current subset.
                query.query();
            }
```